File: 2021/06/06.py
def preprocessing(input): 
    return {n: input.count(str(n)) for n in range(9)}

def solver(lanternfishes):
    for day in range(256):
        if day == 80: yield sum(lanternfishes.values())
        lanternfishes[9]  = lanternfishes[0]
        lanternfishes[7] += lanternfishes[0]
        lanternfishes     = {n: lanternfishes[n + 1] for n in range(9)}
    yield sum(lanternfishes.values())

File: 2021/06/06.rs
aoc::main!(); 

fn preprocessing(input: &str) -> Vec<usize> {
    input
    .split(',')
    .counts()
    .into_iter()
    .sorted()
    .map(|(_, cnt)| cnt)
    .pad_using(9, |_| 0)
    .collect_vec()
}

fn part_1(lanternfishes: Vec<usize>) -> usize {
    get_lf_qty_after_n_days(lanternfishes, 80)
}
 
fn part_2(lanternfishes: Vec<usize>) -> usize {
    get_lf_qty_after_n_days(lanternfishes, 256)
}

fn get_lf_qty_after_n_days(lanternfishes: Vec<usize>, days: usize) -> usize {
    (0..days)
    .fold(lanternfishes, |mut lf, _| {
        lf.rotate_left(1);
        lf[5] += lf[7];
        lf})
    .iter()
    .sum()
}
